igvc_navigation/src/d_lite_planner/main.cpp

```diff
@@ -0,0 +1,245 @@
+/**
+Solves for an optimal path using the D* Lite incremental search algorithm.
+
+D* Lite implementation details can be found in DLitePlanner.h
+*/
+
+#include <nav_msgs/Path.h>
+#include <igvc_msgs/map.h>
+#include <std_msgs/Int32.h>
+#include <sensor_msgs/Image.h>
+#include <geometry_msgs/PoseStamped.h>
+
+#include <ros/ros.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <pcl_ros/point_cloud.h>
+#include <igvc_utils/NodeUtils.hpp>
+#include <tf/transform_datatypes.h>
+
+#include <mutex>
+#include <tuple>
+#include <limits>
+#include <vector>
+#include "igvc_navigation/Graph.h"
+#include "DLitePlanner.h"
+
+std::mutex planning_mutex;
```

Comment:
There are a lot of pretty heavy objects in the global scope here. I'd suggest spending some time to figure out if this file can be refactored with an object or more functions to depend less on this large set of globals.

Reply:
I see these packages as being heavily dependent on one another, yet separate in their functionality and purpose. Trying to squash them together into a single object or a lighter-weight implementation might lead to a decrease in interpret-ability that isn't worth the marginally small performance increase i think this would bring.

Reply:
There is no real performance to be gained or lost here. The suggestion is based on concerns about safety and maintainability. These concerns stem from the difficulty that global variables introduce when it comes to tracking object lifetime and usage.

For example, `expanded_pub` is defined at global scope despite it only being used outside of `main` in `expanded_callback`. I, personally, would prefer that `expanded_pub` was simply declared in `main`, `expanded_callback` renamed to `publish_expanded_set`, and add the publisher as a parameter to the `publish_expanded_set` function. This makes the lifetime and usage of the publisher object more obvious, and the name change reflects the way the function is actually used. It is not a callback function, simply a utility function for converting and publishing the expanded nodes set. 

Similarly, `expanded_cloud` is defined globally despite being used exclusively inside of `expanded_callback`. (The only use in main is actually duplicate code setting the frame_id which `expanded_callback` already sets.) It would be clearer if the cloud was declared inside the function, filled with useful data, and then handed over to the publisher.

Furthermore, many of these variables are declared in global and only actually used in `main`! `CSpace`,`goal_range`,`path_pub`,`rateTime`,`follow_old_path`,`initialize_search`, and `publish_expanded` are all used exclusively in `main`. So moving those variables alone without addressing the other concerns would shrink the list of globals dramatically and make it more obvious to future readers which variables get used in multiple places.

Reply:
Ahh ok I understand your point, and I agree. I'll limit the scope of all variables that shouldn't be in the global scope.